### PR TITLE
1252 additional data

### LINKED ISSUE #1252: Improve UI-Output of additional data field
- related to #1203 
- more user-friendly output
- add the possibility to edit the field via frontend

### PR DESCRIPTION
**Pull request by daniel-arnauer**
### Checklist

- [ ] I have followed (at least) the [PR section of the contributing guide](https://github.com/openkfw/TruBudget/blob/main/CONTRIBUTING.md).
- [ ] I fixed all necessary PR warnings
- [ ] The commit history is clean
- [ ] The E2E tests are passing
- [ ] If possible, the issue has been divided into more subtasks
- [ ] I did a self review before requesting a review from another team member

### Description

Closes #1252

### CODE REVIEW

File: frontend/src/pages/Common/JsonEditor.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    return {
+      json: data
+    };
+  }, [data]);
+
+  const changeDataCallback = useCallback((rawObject) => onChange(rawObject.json), [onChange]);
+
+  useEffect(() => {
+    // create editor
+    // console.log("create editor", refContainer.current);
```
Comment (mayrmartin) [inline frontend/src/pages/Common/JsonEditor.js:21]: pls remove console.log

File: frontend/src/pages/Common/JsonEditor.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        refEditor.current.destroy();
+        refEditor.current = null;
+      }
+    };
+  }, []);
+
+  // update props
+  useEffect(() => {
+    if (refEditor.current) {
+      // console.log("update props", { content: dataObject, onChange: changeDataCallback });
```
Comment (mayrmartin) [inline frontend/src/pages/Common/JsonEditor.js:42]: pls remove console.log

File: frontend/package.json [hunk 3]
```diff
@@ -45,6 +45,7 @@
     "file-saver": "^2.0.5",
     "history": "^4.9.0",
     "immutable": "^3.8.2",
+    "jsoneditor": "^9.9.0",
```
Comment (mayrmartin) [inline frontend/package.json:48]: pls remove one of the two JSON editors, only one is used in your code. 
If possible keep `jsoneditor` (more mature), if you prefer to use a wrapper for the `jsoneditor` have a look at [here](https://www.npmjs.com/package/jsoneditor-react) this is a react wrapper for `jsoneditor`

**Review by mayrmartin — CHANGES_REQUESTED** [review]

File: frontend/src/api.js [hunk 4]
```diff
@@ -144,29 +144,30 @@ class Api {
...
+  declineNode = (node) =>
     instance.post(`/network.declineNode`, {
       node
     });
   listProjects = () => instance.get(`/project.list`);
-  listSubprojects = projectId => instance.get(removeEmptyQueryParams(`/subproject.list?projectId=${projectId}`));
+  listSubprojects = (projectId) => instance.get(removeEmptyQueryParams(`/subproject.list?projectId=${projectId}`));
 
-  createProject = (displayName, description, thumbnail, projectedBudgets, tags) =>
+  createProject = (displayName, description, thumbnail, projectedBudgets, tags, additionalData = { test: "abc" }) =>
```
Comment (mayrmartin) [inline frontend/src/api.js:162]: Pls remove the default value. Additional-data field should be empty if not set

File: scripts/development/docker-compose.yml [hunk 5]
```diff
@@ -211,17 +211,17 @@ services:
       context: ../../frontend
       dockerfile: Dockerfile.dev
       args:
-        API_HOST: alpha-api
+        API_HOST: 172.20.0.4
```
Comment (mayrmartin) [inline scripts/development/docker-compose.yml:214]: For consistency in the project do not change the API_HOST in the dev set-up. Pls reset to `alpha-api`.

**Review by mayrmartin — CHANGES_REQUESTED** [review]

File: frontend/src/sagas.js [hunk 6]
```diff
@@ -597,10 +597,15 @@ export function* editProjectSaga({ projectId, changes, deletedProjectedBudgets =
...
-    if (Object.values(rest).some(value => value !== undefined)) {
+    if (Object.values(rest).some((value) => value !== undefined)) {
       yield callApi(api.editProject, projectId, rest);
     }
 
+    yield put({
+      type: SNACKBAR_MESSAGE,
+      message: "PROJECT CHANGED"
+    });
+
```
Comment (daniel-arnauer) [inline frontend/src/sagas.js:608]: remove or add to language files

File: e2e-test/cypress/integration/project_edit_spec.js [hunk 7]
```diff
@@ -78,4 +78,331 @@ describe("Project Edit", function () {
...
+
+      // Change additional data back to original
+      cy.get(`[data-test=project-additional-data]`).within(() => {
+        cy.get(".jse-value")
+          .click()
+          .type(`${previousAddtionalData}{enter}`);
+      });
+      cy.get(`[data-test=project-additional-data]`).click();
+      cy.get(`[data-test=project-additional-data-submit]`).click();
+      cy.wait(1000);
```
Comment (daniel-arnauer) [inline e2e-test/cypress/integration/project_edit_spec.js:152]: remove cy.wait

File: e2e-test/cypress/integration/project_edit_spec.js [hunk 8]
```diff
@@ -78,4 +78,331 @@ describe("Project Edit", function () {
...
+    cy.grantProjectPermission(projectId, "project.update", executingUser);
+  });
+
+  it("If additional data exists, you can open additional data dialog in list view", function() {
+    cy.login(executingUser, "test");
+    cy.createProject("AdditionalDataTest", undefined, undefined, undefined, {
+      additionalData: {
+        test:
+          "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet."
+      }
```
Comment (daniel-arnauer) [inline e2e-test/cypress/integration/project_edit_spec.js:288]: You can set a shorter test string

File: e2e-test/cypress/integration/project_edit_spec.js [hunk 9]
```diff
@@ -78,4 +78,331 @@ describe("Project Edit", function () {
...
+            previousAddtionalData = additionalData;
+          });
+        cy.get(".jse-value")
+          .click()
+          .type("-changed{enter}");
+      });
+
+      cy.get(`[data-test=project-additional-data]`).click();
+      cy.get(`[data-test=project-additional-data-submit]`).click();
+      cy.wait(1000);
```
Comment (daniel-arnauer) [inline e2e-test/cypress/integration/project_edit_spec.js:371]: .

File: e2e-test/cypress/integration/project_edit_spec.js [hunk 10]
```diff
@@ -78,4 +78,331 @@ describe("Project Edit", function () {
...
+  });
+
+  it("Editing the additional data in the list view is possible", function() {
+    let previousAddtionalData;
+
+    cy.login(executingUser, "test");
+    cy.createProject("AdditionalDataTest", undefined, undefined, undefined, {
+      additionalData: {
+        test:
+          "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet."
```
Comment (daniel-arnauer) [inline e2e-test/cypress/integration/project_edit_spec.js:339]: .

File: e2e-test/cypress/integration/project_edit_spec.js [hunk 11]
```diff
@@ -78,4 +78,331 @@ describe("Project Edit", function () {
...
+    //Changing additional data back to default
+    cy.get(`[data-test=project-additional-data]`).within(() => {
+      cy.get(".jse-value")
+        .click()
+        .type(`${previousAddtionalData}{enter}`);
+    });
+    cy.get(`[data-test=project-additional-data-submit]`)
+      .should("be.visible")
+      .click();
+    cy.wait(1000);
```
Comment (daniel-arnauer) [inline e2e-test/cypress/integration/project_edit_spec.js:394]: .

File: frontend/package.json [hunk 12]
```diff
@@ -65,7 +65,8 @@
     "redux-immutable": "^4.0.0",
     "redux-logger": "^3.0.6",
     "redux-saga": "^1.1.3",
-    "styled-components": "^5.3.5"
+    "styled-components": "^5.3.5",
```
Comment (mayrmartin) [inline frontend/package.json:68]: pls remove styled-components as ui framework; we only use materiual ui (mui5) for styling

**Review by mayrmartin — CHANGES_REQUESTED** [review]
Pls also remove `TruBudget - Verknüpfung.lnk`

File: frontend/src/pages/Overview/TableView.js [hunk 13]
```diff
@@ -43,6 +43,8 @@ const ProjectButtons = ({
   const canViewPermissions = canViewProjectPermissions(allowedIntents);
   const editDisabled = !(canUpdateProject(allowedIntents) && isOpen);
   const viewDisabled = !canViewProjectDetails(allowedIntents);
+  // const additionalDataEmpty = !project.additionalData.test;
```
Comment (mayrmartin) [inline frontend/src/pages/Overview/TableView.js:46]: pls remove comment if not used

**Review by mayrmartin — CHANGES_REQUESTED** [review]

**Review by mayrmartin — APPROVED** [review]

**Pull request merged**